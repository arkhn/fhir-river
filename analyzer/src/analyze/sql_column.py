from .cleaning_script import CleaningScript
from .concept_map import ConceptMap


class SqlColumn:
    def __init__(
        self,
        table: str,
        column: str,
        owner: str = None,
        cleaning_script: CleaningScript = None,
        concept_map: ConceptMap = None,
    ):
        self.table = table.strip()
        self.column = column.strip()
        self.owner = owner.strip() if owner else None

        self.cleaning_script = cleaning_script
        self.concept_map = concept_map

        self.truncated_name = None

    def __eq__(self, other) -> bool:
        return (
            self.table == other.table and self.column == other.column and self.owner == other.owner
        )

    def __str__(self) -> str:
        if self.owner:
            return f"{self.owner}.{self.table}.{self.column}"
        else:
            return f"{self.table}.{self.column}"

    def __hash__(self):
        return hash(str(self))

    def table_name(self) -> str:
        if self.owner:
            return f"{self.owner}.{self.table}"
        else:
            return f"{self.table}"

    def dataframe_column_name(self, index=None):
        """ sqlalchemy builds column names as {table}_{column}.
        This method helps retrieving the needed columns from the dataframe.
        """
        name = f"{self.table}_{self.column}"
        if len(name) <= 30:
            return name

        # Otherwise, we may have problems with sql not accepting aliases of length > 30
        # so we truncate and replace the end by the hex encoded index

        if self.truncated_name is not None:
            return self.truncated_name

        if index is None:
            raise ValueError(
                "Aliases can't be longer than 30 character, please provide an index argument."
            )

        truncated = name[:-5]
        hex_index = format(index, "5x")
        self.truncated_name = f"{truncated}{hex_index}"
        return self.truncated_name
