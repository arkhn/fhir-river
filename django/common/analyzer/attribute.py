import logging
from decimal import Decimal
from typing import List

from .input_group import InputGroup

logger = logging.getLogger(__name__)

numerical_types_map = {
    "integer": int,
    "decimal": float,
    "unsignedInt": int,
    "positiveInt": int,
}


def normalize_to_str(value):
    if isinstance(value, float):
        value = Decimal(value).normalize()
        return format(value, "f")
    return str(value)


class Attribute:
    def __init__(
        self,
        path,
        definition_id: str = None,
        input_groups: List[InputGroup] = None,
    ):
        self.path = path
        self.input_groups = input_groups or []
        self.type = numerical_types_map.get(definition_id, str)
        self.normalizer = numerical_types_map.get(definition_id, normalize_to_str)

    def __eq__(self, other):
        if not isinstance(other, Attribute):
            return False
        return self.path == other.path and self.input_groups == other.input_groups

    def __str__(self):
        return f"path: {self.path}, groups: {self.input_groups}"

    def __hash__(self):
        return hash("{self.path}{self.input_groups}")

    def add_input_group(self, new_group):
        self.input_groups.append(new_group)

    def cast_type(self, value):
        if value is None:
            return None

        try:
            return self.normalizer(value)
        except Exception as e:
            logger.warning(
                f"Could not cast value {value} to type {self.type} " f"on attribute at path = {self.path}): {e}"
            )

        return value
