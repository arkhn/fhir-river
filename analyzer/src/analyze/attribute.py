from typing import List

from analyzer.src.config.logger import create_logger

from .sql_column import SqlColumn
from .merging_script import MergingScript

logger = create_logger("attribute")

numerical_types_map = {
    "integer": int,
    "decimal": float,
    "unsignedInt": int,
    "positiveInt": int,
}


class Attribute:
    def __init__(
        self,
        path,
        definition_id: str = None,
        columns: List[SqlColumn] = None,
        static_inputs: List[str] = None,
        merging_script: MergingScript = None,
    ):
        self.path = path
        self.type = numerical_types_map.get(definition_id, str)
        self.columns = columns or []
        self.static_inputs = static_inputs or []
        self.merging_script = merging_script

    def __eq__(self, other):
        if not isinstance(other, Attribute):
            return False
        return (
            self.path == other.path
            and self.columns == other.columns
            and self.static_inputs == other.static_inputs
            and self.merging_script == other.merging_script
        )

    def __str__(self):
        return (
            f"path: {self.path}, columns: {self.columns}, "
            f"static_inputs: {self.static_inputs}, merging_script: {self.merging_script}"
        )

    def __hash__(self):
        return hash("{self.path}{self.columns}{self.static_inputs}{self.merging_script}")

    def add_column(self, col):
        self.columns.append(col)

    def add_static_input(self, value):
        self.static_inputs.append(self.cast_type(value))

    def cast_type(self, value):
        if value is None:
            return None

        try:
            return self.type(value)
        except:
            logger.error(
                f"Could not cast value {value} to type {self.type} "
                f"on attribute at path = {self.path}): {e}"
            )
