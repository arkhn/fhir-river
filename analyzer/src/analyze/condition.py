from typing import Dict, List, NewType, Tuple, Union

from .sql_column import SqlColumn

CONDITION_FLAG = "__condition__"


DataDictKey = NewType("DataDictKey", Tuple[str, Tuple[str, str]])
DataDictValue = NewType("DataDictValue", Union[str, List[str]])


class Condition:
    def __init__(
        self, action: str = None, sql_column: SqlColumn = None, value: str = None,
    ):
        self.action = action
        self.sql_column = sql_column
        self.value = value

    def check(self, data: Dict[DataDictKey, DataDictValue]):
        data_value = data[(CONDITION_FLAG, (self.sql_column.table, self.sql_column.column))]

        # if data_value is an array, all of its values should be similar
        if isinstance(data_value, tuple):
            if not all(el == data_value[0] for el in data_value[1:]):
                raise ValueError(
                    "Conditions can only be checked against arrays with identical values, "
                    f"got {data_value}."
                )
            data_value = data_value[0]

        # TODO handle types
        return (self.action == "EXCLUDE") ^ (self.value == str(data_value))
