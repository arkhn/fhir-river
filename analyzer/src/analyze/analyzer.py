import re
import time
from collections.abc import Mapping

from analyzer.src.analyze.graphql import PyrogClient
from analyzer.src.config.logger import get_logger

from .analysis import Analysis
from .attribute import Attribute
from .cleaning_script import CleaningScript
from .concept_map import ConceptMap
from .condition import Condition
from .input_group import InputGroup
from .mapping import build_squash_rules
from .merging_script import MergingScript
from .sql_column import SqlColumn
from .sql_join import SqlJoin
from .sql_filter import SqlFilter

logger = get_logger()


class Analyzer:
    def __init__(self, pyrog_client: PyrogClient):
        self.pyrog = pyrog_client
        # Store analyses
        # TODO think about the design here. Use http caching instead of
        # storing here, for instance?
        self.analyses: Mapping = {}
        self.last_updated_at: Mapping = {}  # store last updated timestamp for each resource_id

        self._cur_analysis = Analysis()

    def get_analysis(self, resource_mapping_id, max_seconds_refresh=3600) -> Analysis:
        if resource_mapping_id not in self.analyses:
            self.fetch_analysis(resource_mapping_id)
        else:
            self.check_refresh_analysis(resource_mapping_id, max_seconds_refresh)
        return self.analyses[resource_mapping_id]

    def check_refresh_analysis(self, resource_mapping_id, max_seconds_refresh):
        """
        This method refreshes the analyser if the last update was later than `max_seconds_refresh`
        for each resource
        """
        if time.time() - self.last_updated_at.get(resource_mapping_id) > max_seconds_refresh:
            logger.info(
                f"Analysis too old for resource {resource_mapping_id}.",
                extra={"resource_id": resource_mapping_id},
            )
            self.fetch_analysis(resource_mapping_id)
        else:
            logger.info(
                "Analysis was updated recently. Using cached analysis.",
                extra={"resource_id": resource_mapping_id},
            )

    def fetch_analysis(self, resource_mapping_id):
        """
        Fetch mapping from API and store last updated timestamp
        :param resource_mapping_id:
        :return:
        """
        logger.info("Fetching mapping from api.", extra={"resource_id": resource_mapping_id})
        resource_mapping = self.pyrog.get_resource_from_id(resource_id=resource_mapping_id)
        self.analyze(resource_mapping)
        self.last_updated_at[resource_mapping_id] = time.time()

    # TODO add an update_analysis(self, resource_mapping_id)?

    def analyze(self, resource_mapping):
        self._cur_analysis = Analysis()

        # Analyze the mapping
        self.analyze_mapping(resource_mapping)

        if not self._cur_analysis.columns:
            self._cur_analysis.is_static = True
        else:
            # Add primary key to columns to fetch if needed
            self._cur_analysis.add_column(self._cur_analysis.primary_key_column)

            # Build squash rules
            self._cur_analysis.squash_rules = build_squash_rules(
                self._cur_analysis.columns,
                self._cur_analysis.joins,
                self._cur_analysis.primary_key_column.table_name(),
            )

        # Store analysis
        self.analyses[resource_mapping["id"]] = self._cur_analysis

        return self._cur_analysis

    def analyze_mapping(self, resource_mapping):
        self._cur_analysis.primary_key_column = self.get_primary_key(resource_mapping)
        self._cur_analysis.source_id = resource_mapping["source"]["id"]
        self._cur_analysis.source_credentials = resource_mapping["source"]["credential"]
        self._cur_analysis.resource_id = resource_mapping["id"]
        self._cur_analysis.definition_id = resource_mapping["definitionId"]
        self._cur_analysis.definition = resource_mapping["definition"]
        for filter_ in resource_mapping["filters"]:
            self.analyze_filter(filter_)
        for attribute_mapping in resource_mapping["attributes"]:
            self.analyze_attribute(attribute_mapping)

        return self._cur_analysis

    def analyze_filter(self, filter_):
        col = SqlColumn(
            filter_["sqlColumn"]["table"],
            filter_["sqlColumn"]["column"],
            self._cur_analysis.source_credentials["owner"],
        )
        sql_filter = SqlFilter(col, filter_["relation"], filter_["value"])
        self._cur_analysis.add_filter(sql_filter)

    def analyze_attribute(self, attribute_mapping):
        logger.info(
            f"Analyze attribute {attribute_mapping['path']} {attribute_mapping['definitionId']}",
            extra={"resource_id": self._cur_analysis.resource_id},
        )
        attribute = Attribute(
            path=attribute_mapping["path"], definition_id=attribute_mapping["definitionId"]
        )
        if not attribute_mapping["inputGroups"]:
            # If there are no input groups for this attribute, it means that it is an intermediary
            # attribute (ie not a leaf). It is here to give us some context information.
            # For instance, we can use it if its children attributes represent a Reference.
            if attribute_mapping["definitionId"] == "Reference":
                logger.info(
                    f"Analyze attribute reference",
                    extra={"resource_id": self._cur_analysis.resource_id},
                )
                # Remove trailing index
                path = re.sub(r"\[\d+\]$", "", attribute.path)
                self._cur_analysis.reference_paths.add(path)

            return

        for mapping_group in attribute_mapping["inputGroups"]:
            self.analyze_input_group(mapping_group, attribute)

        self._cur_analysis.attributes.append(attribute)

        return attribute

    def analyze_input_group(self, mapping_group, parent_attribute):
        input_group = InputGroup(id_=mapping_group["id"], attribute=parent_attribute)
        parent_attribute.add_input_group(input_group)
        for input_ in mapping_group["inputs"]:
            if input_["sqlValue"]:

                sqlValue = input_["sqlValue"]
                cur_col = SqlColumn(
                    sqlValue["table"],
                    sqlValue["column"],
                    self._cur_analysis.primary_key_column.owner,
                )

                if input_["script"]:
                    cur_col.cleaning_script = CleaningScript(input_["script"])

                if input_["conceptMapId"]:
                    cur_col.concept_map = ConceptMap(input_["conceptMapId"])

                for join in sqlValue["joins"]:
                    tables = join["tables"]
                    left = SqlColumn(
                        tables[0]["table"],
                        tables[0]["column"],
                        self._cur_analysis.primary_key_column.owner,
                    )
                    right = SqlColumn(
                        tables[1]["table"],
                        tables[1]["column"],
                        self._cur_analysis.primary_key_column.owner,
                    )
                    self._cur_analysis.add_join(SqlJoin(left, right))

                self._cur_analysis.add_column(cur_col)
                input_group.add_column(cur_col)

            elif input_["staticValue"]:
                input_group.add_static_input(input_["staticValue"])

        for mapping_condition in mapping_group["conditions"]:
            condition_column = SqlColumn(
                mapping_condition["sqlValue"]["table"],
                mapping_condition["sqlValue"]["column"],
                self._cur_analysis.primary_key_column.owner,
            )
            self._cur_analysis.add_column(condition_column)

            condition = Condition(
                action=mapping_condition["action"],
                sql_column=condition_column,
                value=mapping_condition["value"],
            )
            input_group.add_condition(condition)

        if mapping_group["mergingScript"]:
            input_group.merging_script = MergingScript(mapping_group["mergingScript"])

        return input_group

    def get_primary_key(self, resource_mapping):
        """ Get the primary key table and column of the provided resource.
        """
        if not resource_mapping["primaryKeyTable"] or not resource_mapping["primaryKeyColumn"]:
            raise ValueError(
                "You need to provide a primary key table and column in the mapping for "
                f"resource {resource_mapping['definitionId']}."
            )

        return SqlColumn(
            resource_mapping["primaryKeyTable"],
            resource_mapping["primaryKeyColumn"],
            resource_mapping["source"]["credential"]["owner"],
        )
