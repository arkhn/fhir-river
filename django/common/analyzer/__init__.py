from .analyzer import Analyzer  # noqa
from .concept_map import ConceptMap  # noqa
