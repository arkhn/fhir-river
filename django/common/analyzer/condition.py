from datetime import datetime
from typing import Dict, List, NewType, Tuple, Union

from common.analyzer.errors import OperationOutcome
from common.analyzer.sql_column import SqlColumn

CONDITION_FLAG = "__condition__"

# NOTE If the reference value is null, we want the relation to be False,
# that's why we check for `x is not None` in all the binary relations.
# For instance, if we want to INCLUDE the attribute if a column EQ a value and
# this value is null, we won't keep it. On the other hand,
# if we want to EXCLUDE the attribute if a column EQ a value and
# this value is null, we will keep it.
CONDITION_RELATION_TO_FUNCTION = {
    "EQ": lambda x, y: x is not None and x == y,
    "NEQ": lambda x, y: x is not None and x != y,
    "LT": lambda x, y: x is not None and x < y,
    "LE": lambda x, y: x is not None and x <= y,
    "GE": lambda x, y: x is not None and x >= y,
    "GT": lambda x, y: x is not None and x > y,
    "NULL": lambda x, _: x is None,
    "NOTNULL": lambda x, _: x is not None,
}

UNARY_RELATIONS = ["NULL", "NOTNULL"]

DataDictKey = NewType("DataDictKey", Tuple[str, Tuple[str, str]])
DataDictValue = NewType("DataDictValue", Union[str, List[str]])


class Condition:
    def __init__(self, action: str = None, sql_column: SqlColumn = None, relation: str = None, value: str = None):
        self.action = action
        self.sql_column = sql_column
        self.relation = relation
        # We turn the value into a number if it looks like one
        self.value = value

    def check(self, data: Dict[DataDictKey, DataDictValue]):
        data_value = data[(CONDITION_FLAG, (self.sql_column.table_name(), self.sql_column.column))]

        try:
            cast_value = self.cast_value_type(data_value[0])
        except Exception as e:
            raise OperationOutcome(f"Could not cast condition value ({self.value}) to type {type(data_value[0])}: {e}")

        results = []
        for value in data_value:
            # We first check if the relation between the condition's value and
            # the value fetched from the DB holds.
            is_relation_true = CONDITION_RELATION_TO_FUNCTION[self.relation](value, cast_value)

            # Then, to know if we need to include the input group or not, we need to XOR
            # is_relation_true with self.action == "EXCLUDE".
            # For instance, if the relation holds but the action is "EXCLUDE", we want
            # to return False (and to exclude the input group from the attribute).
            results.append((self.action == "EXCLUDE") ^ is_relation_true)

        return results

    def cast_value_type(self, data_value):
        if self.relation in UNARY_RELATIONS or data_value is None:
            # For unary relations, we don't need a reference value
            cast_value = None
        elif isinstance(data_value, bool):
            # For booleans, we cast to False if the value is "0", "false" or "False"
            cast_value = self.value.lower() not in ["0", "false"]
        elif isinstance(data_value, str) and self.is_date(data_value):
            # For dates, we parse the string
            # NOTE the input date format is fixed here
            cast_value = datetime.strptime(self.value, "%Y-%m-%d").strftime("%Y-%m-%dT%H:%M:%S")
        else:
            cast_value = type(data_value)(self.value)

        return cast_value

    @staticmethod
    def is_date(raw_input):
        try:
            # flask jsonifies dates to this format
            datetime.strptime(raw_input, "%Y-%m-%dT%H:%M:%S")
            return True
        except ValueError:
            return False

    def __eq__(self, other):
        if not isinstance(other, Condition):
            return False
        return (
            self.action == other.action
            and self.sql_column == other.sql_column
            and self.relation == other.relation
            and self.value == other.value
        )
