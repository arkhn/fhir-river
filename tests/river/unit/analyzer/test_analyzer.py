from uuid import uuid4

from river.common.analyzer import Analyzer
from river.common.analyzer.attribute import Attribute
from river.common.analyzer.condition import Condition
from river.common.analyzer.input_group import InputGroup
from river.common.analyzer.sql_column import SqlColumn
from river.common.analyzer.sql_join import SqlJoin


def test_cache_analysis_redis(mimic_mapping):
    batch_id = uuid4()
    resource_id = "cktiqt0c800tfxgutkq5b6gct"
    analyzer = Analyzer()

    res = analyzer.cache_analysis(batch_id, resource_id, mimic_mapping)

    assert res.definition_id == "Patient"
    assert res.primary_key_column.table == "patients"
    assert res.primary_key_column.column == "subject_id"

    assert analyzer.analyses[f"{batch_id}:{resource_id}"].definition_id == "Patient"
    assert analyzer.analyses[f"{batch_id}:{resource_id}"].primary_key_column.table == "patients"
    assert analyzer.analyses[f"{batch_id}:{resource_id}"].primary_key_column.column == "subject_id"

    cached_res = analyzer.load_analysis(batch_id, resource_id)

    assert cached_res == res


def test_get_primary_key():
    analyzer = Analyzer()

    resource_mapping = {
        "primary_key_owner": "owner_id",
        "primary_key_table": "table",
        "primary_key_column": "col",
    }
    analyzer._owners_data = {"owner_id": "owner"}

    assert analyzer.get_primary_key(resource_mapping) == SqlColumn("owner", "table", "col")


def test_get_primary_key_missing_field():
    analyzer = Analyzer()

    resource_mapping = {
        "primary_key_owner": "owner_id",
        "primary_key_table": "",
        "primary_key_column": "col",
    }
    analyzer._owners_data = {"owner_id": "owner"}

    assert analyzer.get_primary_key(resource_mapping) is None


def test_analyze_mapping(mimic_mapping, snapshot):
    analyzer = Analyzer()
    resource_id = "cktiqszf200aexgutj9ttg8l1"

    analysis = analyzer.analyze(resource_id, mimic_mapping)

    assert len(analysis.attributes) == 14
    assert analysis == snapshot
    assert analyzer.get_analysis_columns(analysis) == snapshot
    assert analysis.filters == snapshot
    assert analysis.reference_paths == [["partOf"], ["subject"], ["encounter"]]


def test_analyze_attribute(dict_map_gender, structure_definitions):
    analyzer = Analyzer()
    analyzer._cur_analysis.primary_key_column = SqlColumn("mimiciii", "patients", "subject_id")
    analyzer._cur_analysis.definition_id = "Patient"
    analyzer._cur_analysis.definition = next(
        iter(
            [
                structure_definition
                for structure_definition in structure_definitions
                if structure_definition["id"] == analyzer._cur_analysis.definition_id
            ]
        )
    )
    analyzer._columns_data = {
        "ck8ooenw827004kp41nv3kcmq": {
            "owner": "mimiciii",
            "table": "patients",
            "column": "gender",
        },
        "ckdyl65kj0196gu9ku2dy0ygg": {
            "owner": "mimiciii",
            "table": "patients",
            "column": "subject_id",
        },
        "ckdyl65kj0197gu9k1lrvx3bl": {
            "owner": "mimiciii",
            "table": "admissions",
            "column": "subject_id",
        },
        "ckdyl65kl0335gu9kup0hwhe0": {
            "owner": "mimiciii",
            "table": "admissions",
            "column": "expire_flag",
        },
        "ckdyl65kj0196gu9ku2dy0ygb": {
            "owner": "mimiciii",
            "table": "patients",
            "column": "subject_id",
        },
        "ckdyl65kj0197gu9k1lrvx3bb": {
            "owner": "mimiciii",
            "table": "join_table",
            "column": "subject_id",
        },
        "ckdyl65kj0196gu9ku2dy0yga": {
            "owner": "mimiciii",
            "table": "join_table",
            "column": "adm_id",
        },
        "ckdyl65kj0197gu9k1lrvx3ba": {
            "owner": "mimiciii",
            "table": "admissions",
            "column": "adm_id",
        },
    }
    attribute_mapping = {
        "id": "ck8ooenpu26984kp4wyiz4yc2",
        "path": "gender",
        "definition_id": "code",
        "resource_id": "ck8oo3on226974kp4ns32n7xs",
        "comments": [],
        "input_groups": [
            {
                "id": "ckdom8lgq0045m29ksz6vudvc",
                "merging_script": None,
                "static_inputs": [],
                "sql_inputs": [
                    {
                        "script": "",
                        "concept_map_id": "id_cm_gender",
                        "concept_map": dict_map_gender,
                        "column": "ck8ooenw827004kp41nv3kcmq",
                        "joins": [{"left": "ckdyl65kj0196gu9ku2dy0ygg", "right": "ckdyl65kj0197gu9k1lrvx3bl"}],
                    }
                ],
                "conditions": [
                    {
                        "action": "EXCLUDE",
                        "relation": "EQ",
                        "value": "1",
                        "sql_input": {
                            "script": "",
                            "concept_map_id": "",
                            "column": "ckdyl65kl0335gu9kup0hwhe0",
                            "joins": [
                                {"left": "ckdyl65kj0196gu9ku2dy0ygb", "right": "ckdyl65kj0197gu9k1lrvx3bb"},
                                {"left": "ckdyl65kj0196gu9ku2dy0yga", "right": "ckdyl65kj0197gu9k1lrvx3ba"},
                            ],
                        },
                    }
                ],
            }
        ],
    }

    actual = analyzer.analyze_attribute(attribute_mapping)

    expected = Attribute("gender")

    group = InputGroup(
        id_="ckdom8lgq0045m29ksz6vudvc",
        attribute=expected,
        conditions=[
            Condition(
                "EXCLUDE",
                SqlColumn(
                    "mimiciii",
                    "admissions",
                    "expire_flag",
                    joins=[
                        SqlJoin(
                            SqlColumn("mimiciii", "patients", "subject_id"),
                            SqlColumn("mimiciii", "join_table", "subject_id"),
                        ),
                        SqlJoin(
                            SqlColumn("mimiciii", "join_table", "adm_id"),
                            SqlColumn("mimiciii", "admissions", "adm_id"),
                        ),
                    ],
                ),
                "EQ",
                "1",
            )
        ],
        columns=[
            SqlColumn(
                "mimiciii",
                "patients",
                "gender",
                joins=[
                    SqlJoin(
                        SqlColumn("mimiciii", "patients", "subject_id"),
                        SqlColumn("mimiciii", "admissions", "subject_id"),
                    )
                ],
            )
        ],
        static_inputs=[],
        merging_script=None,
    )
    expected.add_input_group(group)

    assert actual == expected
