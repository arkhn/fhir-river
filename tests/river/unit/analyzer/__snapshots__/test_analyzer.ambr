# name: test_analyze_mapping
  <class 'Analysis'> {
    attributes=<class 'list'> [
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszf700ahxgut1amf3sb3',
            merging_script=None,
            static_inputs=<class 'list'> [
              'unknown',
            ],
          },
        ],
        path='status',
        type='code',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszfl00apxgut3efwipse',
            merging_script=None,
            static_inputs=<class 'list'> [
              'http ://loinc.org',
            ],
          },
        ],
        path='code.coding[0].system',
        type='uri',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszfu00auxgutgcraw8u2',
            merging_script=None,
            static_inputs=<class 'list'> [
              '29463-7',
            ],
          },
        ],
        path='code.coding[0].code',
        type='code',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
              <class 'SqlColumn'> {
                cleaning_script=<Script: clean_dateTime>,
                column='dischtime',
                concept_map=None,
                joins=<class 'list'> [
                ],
                owner='public',
                table='admissions',
              },
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszfy00axxgut06tymipu',
            merging_script=None,
            static_inputs=<class 'list'> [
              'Body weight',
            ],
          },
        ],
        path='code.coding[0].display',
        type='string',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszg300b0xgutcctthoa3',
            merging_script=None,
            static_inputs=<class 'list'> [
              'Patient',
            ],
          },
        ],
        path='subject.type',
        type='uri',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszga00b4xguthp4a16v5',
            merging_script=None,
            static_inputs=<class 'list'> [
              'http://terminology.arkhn.org/038b7a11-2195-4b6d-b5d3-40e5422d7446',
            ],
          },
        ],
        path='subject.identifier.system',
        type='uri',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
              <class 'SqlColumn'> {
                cleaning_script=<Script: clean_dateTime>,
                column='admittime',
                concept_map=None,
                joins=<class 'list'> [
                ],
                owner='public',
                table='chartevents',
              },
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszgh00b8xgutz98bej79',
            merging_script=None,
            static_inputs=<class 'list'> [
            ],
          },
        ],
        path='subject.identifier.value',
        type='string',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszgn00bbxgutfyq60u1n',
            merging_script=None,
            static_inputs=<class 'list'> [
              'Encounter',
            ],
          },
        ],
        path='encounter.type',
        type='uri',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszgu00bfxgut336ul1fr',
            merging_script=None,
            static_inputs=<class 'list'> [
              'http://terminology.arkhn.org/a31b9d5c-6635-4861-8162-577bd26eb010',
            ],
          },
        ],
        path='encounter.identifier.system',
        type='uri',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
              <class 'SqlColumn'> {
                cleaning_script=None,
                column='icustay_id',
                concept_map=None,
                joins=<class 'list'> [
                ],
                owner='public',
                table='chartevents',
              },
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszh100bjxgut3sew1u2z',
            merging_script=None,
            static_inputs=<class 'list'> [
            ],
          },
        ],
        path='encounter.identifier.value',
        type='string',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
              <class 'SqlColumn'> {
                cleaning_script=None,
                column='charttime',
                concept_map=None,
                joins=<class 'list'> [
                ],
                owner='public',
                table='chartevents',
              },
            ],
            conditions=<class 'list'> [
            ],
            id='cktiqszh600bmxgutabgdcab4',
            merging_script=None,
            static_inputs=<class 'list'> [
            ],
          },
        ],
        path='effectiveDateTime',
        type='dateTime',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
            ],
            conditions=<class 'list'> [
              <class 'Condition'> {
                action='EXCLUDE',
                relation='NULL',
                sql_column=<class 'SqlColumn'> {
                  cleaning_script=None,
                  column='valuenum',
                  concept_map=None,
                  joins=<class 'list'> [
                  ],
                  owner='public',
                  table='chartevents',
                },
                value='',
              },
            ],
            id='cktiqszhb00bpxgutpj8j01du',
            merging_script=None,
            static_inputs=<class 'list'> [
              'http://unitsofmeasure.org',
            ],
          },
        ],
        path='valueQuantity.system',
        type='uri',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
            ],
            conditions=<class 'list'> [
              <class 'Condition'> {
                action='EXCLUDE',
                relation='NULL',
                sql_column=<class 'SqlColumn'> {
                  cleaning_script=None,
                  column='valuenum',
                  concept_map=None,
                  joins=<class 'list'> [
                  ],
                  owner='public',
                  table='chartevents',
                },
                value='',
              },
            ],
            id='cktiqszhi00buxgutz3o4iyl7',
            merging_script=None,
            static_inputs=<class 'list'> [
              'kg',
            ],
          },
        ],
        path='valueQuantity.unit',
        type='string',
      },
      <class 'Attribute'> {
        input_groups=<class 'list'> [
          <class 'InputGroup'> {
            attribute=...,
            columns=<class 'list'> [
              <class 'SqlColumn'> {
                cleaning_script=None,
                column='valuenum',
                concept_map=None,
                joins=<class 'list'> [
                ],
                owner='public',
                table='chartevents',
              },
            ],
            conditions=<class 'list'> [
              <class 'Condition'> {
                action='EXCLUDE',
                relation='NULL',
                sql_column=<class 'SqlColumn'> {
                  cleaning_script=None,
                  column='valuenum',
                  concept_map=None,
                  joins=<class 'list'> [
                  ],
                  owner='public',
                  table='chartevents',
                },
                value='',
              },
            ],
            id='cktiqszhp00byxguti7n2evk5',
            merging_script=None,
            static_inputs=<class 'list'> [
            ],
          },
        ],
        path='valueQuantity.value',
        type='decimal',
      },
    ],
    definition=None,
    definition_id='Observation',
    filters=<class 'list'> [
      <class 'SqlFilter'> {
        relation='IN',
        sql_column=<class 'SqlColumn'> {
          cleaning_script=None,
          column='itemid',
          concept_map=None,
          joins=<class 'list'> [
          ],
          owner='public',
          table='chartevents',
        },
        value='762,3580,3696,226512',
      },
    ],
    is_static=False,
    label='Body_Weight',
    logical_reference='6f691f4c-920b-4cc0-b241-67f77f43850e',
    primary_key_column=<class 'SqlColumn'> {
      cleaning_script=None,
      column='row_id',
      concept_map=None,
      joins=<class 'list'> [
      ],
      owner='public',
      table='chartevents',
    },
    reference_paths=<class 'list'> [
      <class 'list'> [
        'partOf',
      ],
      <class 'list'> [
        'subject',
      ],
      <class 'list'> [
        'encounter',
      ],
    ],
    resource_id='cktiqszf200aexgutj9ttg8l1',
    source_credentials=<class 'dict'> {
      'database': 'test',
      'host': 'mimic',
      'login': 'test',
      'model': 'POSTGRES',
      'password': 'test',
      'port': 5432,
    },
    source_id='cktiqsyuj0002xgutzhqbn1ut',
  }
---
# name: test_analyze_mapping.1
  <class 'set'> {
    <class 'SqlColumn'> {
      cleaning_script=None,
      column='charttime',
      concept_map=None,
      joins=<class 'list'> [
      ],
      owner='public',
      table='chartevents',
    },
    <class 'SqlColumn'> {
      cleaning_script=None,
      column='icustay_id',
      concept_map=None,
      joins=<class 'list'> [
      ],
      owner='public',
      table='chartevents',
    },
    <class 'SqlColumn'> {
      cleaning_script=None,
      column='itemid',
      concept_map=None,
      joins=<class 'list'> [
      ],
      owner='public',
      table='chartevents',
    },
    <class 'SqlColumn'> {
      cleaning_script=None,
      column='subject_id',
      concept_map=None,
      joins=<class 'list'> [
      ],
      owner='public',
      table='chartevents',
    },
    <class 'SqlColumn'> {
      cleaning_script=None,
      column='valuenum',
      concept_map=None,
      joins=<class 'list'> [
      ],
      owner='public',
      table='chartevents',
    },
  }
---
# name: test_analyze_mapping.2
  <class 'list'> [
    <class 'SqlFilter'> {
      relation='IN',
      sql_column=<class 'SqlColumn'> {
        cleaning_script=None,
        column='itemid',
        concept_map=None,
        joins=<class 'list'> [
        ],
        owner='public',
        table='chartevents',
      },
      value='762,3580,3696,226512',
    },
  ]
---
